components/support/migration/src/main/java/mozilla/components/support/migration/GeckoMigration.kt

```diff
@@ -17,27 +94,73 @@ internal object GeckoMigration {
      * Tries to migrate Gecko related internal files in [profilePath].
      */
     fun migrate(
-        profilePath: String
-    ): Result<Unit> {
+        profilePath: String,
+        migrationVersion: Int
+    ): Result<GeckoMigrationResult> {
         // GeckoView will happily pick up the profile from Fennec and reuse all data in it. So this
-        // migration is mostly focused on removing the files that we do not want to reuse.
+        // migration is mostly focused on removing all prefs that we do not want to reuse.
+        val prefsjs = File(profilePath, PREFS_FILE)
+        if (!prefsjs.exists()) {
+            return Result.Success(GeckoMigrationResult.Success.NoPrefsFile)
+        }
+
+        // First, let's create a backup of the file.
+        try {
+            prefsjs.copyTo(File(profilePath, PREFS_BACKUP_FILE + migrationVersion), true)
+        } catch (e: IOException) {
+            return Result.Failure(GeckoMigrationException(GeckoMigrationResult.Failure.FailedToWriteBackup(e)))
+        }
 
-        // The prefs file contains gecko prefs that have been changed by the user (about:config) or
-        // code. Since those may be incompatible with GeckoView and/or app code, we decided to just
-        // reset them. GeckoView will recreate the file with defaults if needed.
-        return removePrefsFile(File(profilePath))
+        // Now let's transform the original to contain only prefs we want to keep.
+        val transformed = prefsjs.useLines {
+            it.filter { line ->
+                userPrefsToKeep.any { pref -> line.startsWith("$USER_PREF(\"$pref\"") }
+            }.toList()
+        }
+
+        return if (transformed.isEmpty()) {
+            removePrefsFile(prefsjs)
+        } else {
```

Comment:
style nit: you don't _need_ to wrap this block in an `else` if you `return` inside of the `if`. that will help flatten this code a bit. the cost is repeating `return` twice, though.

Reply:
True, just tried, but to me at least, the single return is easier to read, but that's definitely subjective 🤷‍♂️ .